### ISSUE TITLE
DotNetNuke.Modules.Announcements.Components.Business.AnnouncementInfo.GetProperty Error

### ISSUE THREAD

**Issue by ej401** [DNNCommunity/DNN.Announcements#17]
## Describe the bug
We are seeing this error in the exception log when opening the updated module in page edit mode. 

Note: we also had to add the {databaseOwner}{objectQualifier} string to 2 areas in the 07.00.00.SqlDataProvider in order to get the module to install.

## Software Versions
- DNN: 9.2.0
- Module: 7.2.0

## To Reproduce
Open Announcements module in page edit mode

StackTrace:

InnerMessage:Object reference not set to an instance of an object.

InnerStackTrace:

at DotNetNuke.Modules.Announcements.Components.Business.AnnouncementInfo.GetProperty(String strPropertyName, String strFormat, CultureInfo formatProvider, UserInfo AccessingUser, Scope AccessLevel, Boolean& PropertyNotFound) at DotNetNuke.Services.Tokens.BaseCustomTokenReplace.replacedTokenValue(String objectName, String propertyName, String format) at DotNetNuke.Services.Tokens.BaseTokenReplace.ReplaceTokens(String sourceText) at DotNetNuke.Modules.Announcements.Components.Template.SimpleTemplate.RenderTemplate(IEnumerable`1 announcements, Boolean editEnabled) at DotNetNuke.Modules.Announcements.MVP.Presenters.AnnouncementsViewPresenter.GetRenderedTemplate(Object sender, EventArgs e) at DotNetNuke.Modules.Announcements.AnnouncementsView.OnLoad(EventArgs e) at System.Web.UI.Control.LoadRecursive() at System.Web.UI.Control.LoadRecursive() at System.Web.UI.Control.LoadRecursive() at System.Web.UI.Control.LoadRecursive() at System.Web.UI.Control.LoadRecursive() at System.Web.UI.Control.LoadRecursive() at System.Web.UI.Control.LoadRecursive() at System.Web.UI.Control.LoadRecursive() at System.Web.UI.Control.LoadRecursive() at System.Web.UI.Control.LoadRecursive() at System.Web.UI.Control.LoadRecursive() at System.Web.UI.Control.LoadRecursive() at System.Web.UI.Control.LoadRecursive() at System.Web.UI.Page.ProcessRequestMain(Boolean includeStagesBeforeAsyncPoint, Boolean includeStagesAfterAsyncPoint)

Thanks,
Eric

**Comment by valadas** [comment]
Thanks for the bug report @ej401 , I can confirm the first issue with the object qualifier missing in the sql script. That script will only run on upgrades, some more testing needs to be done regarding upgrades on this module.

For the second issue, I think it may be a problem with the template configuration, can you please grab a screenshot of your template (or copy paste it), thanks.

**Comment by ej401** [comment]
Daniel,

Here is the content of template.xml (it’s the same in the ModernAnnouncements and SimpleAnnouncements directories):

<?xml version="1.0" encoding="utf-8" ?>
<TemplateInfo>
  <Type>SimpleTemplate</Type>
  <ItemTemplate>item.htm</ItemTemplate>
  <AltItemTemplate></AltItemTemplate>
  <HeaderTemplate>header.htm</HeaderTemplate>
  <FooterTemplate>footer.htm</FooterTemplate>
  <SeparatorTemplate></SeparatorTemplate>
  <DetailTemplate></DetailTemplate>
  <CssFile></CssFile>
  <JsFile></JsFile>
  <Version>01.00.00</Version>
  <LastModifiedDate>2012-11-28T23:30:00</LastModifiedDate>
</TemplateInfo>

Thanks,
Eric

**Comment by valadas** [comment]
Hmm, I am not sure how those template files are supposed to be selected, from what I see, the template is directly entered in the module "Template Configuration" menu like this:
![image](https://user-images.githubusercontent.com/6371568/42262118-436907c8-7f38-11e8-8c96-f540f9699749.png)

I think those files where there because they where working on a new feature back when this was in codeplex and that feature was never completed (a new templating engine).

Can you please post a screenshot or copy-paste from that module dialog, there may be an invalid token in there I guess.

**Comment by valadas** [comment]
Or, try loading the default for each textbox and report back

**Comment by ej401** [comment]
Daniel,

I can’t edit the module to even get to the template configuration – that is where the error occurs. I looked in the ModuleSettings table, though, and the template records appear to be in order.

I think the problem is that the new version cannot find the templates of existing announcement module instances – that is the reason for the null object reference. When I add a new announcements module instance, everything works fine.

Has the location of the settings moved? The reason I ask, is because when I created a new module instance, nothing was added to the ModuleSettings table.

I also cleared the cache during this testing.

Thanks,
Eric

**Comment by ej401** [comment]
Daniel,

Ignore what I said below – I found the issue. The “token replacement” references in the stack trace got me to thinking about how I’ve had to deal with NULL database values in some of our .NET apps. The “ImageSource” field in the Announcements table was NULL for several of our existing announcements, but adding a new announcement inserts an empty string for this field if you don’t use an image – this is why the new instance I mentioned below didn’t cause the error. I imagine a code change was made that assumes this value will never be null.

In the short term, we can fix this by simply updating our existing announcement records.

Thanks for your help,
Eric

From: Johnson, Eric
Sent: Thursday, July 5, 2018 9:15 AM
To: 'DNNCommunity/DNN.Announcements' <reply@reply.github.com>; DNNCommunity/DNN.Announcements <DNN.Announcements@noreply.github.com>
Cc: Mention <mention@noreply.github.com>
Subject: RE: [DNNCommunity/DNN.Announcements] DotNetNuke.Modules.Announcements.Components.Business.AnnouncementInfo.GetProperty Error (#17)

Daniel,

I can’t edit the module to even get to the template configuration – that is where the error occurs. I looked in the ModuleSettings table, though, and the template records appear to be in order.

I think the problem is that the new version cannot find the templates of existing announcement module instances – that is the reason for the null object reference. When I add a new announcements module instance, everything works fine.

Has the location of the settings moved? The reason I ask, is because when I created a new module instance, nothing was added to the ModuleSettings table.

I also cleared the cache during this testing.

Thanks,
Eric

**Comment by ej401** [comment]
Daniel,

Something I just noticed when trying to edit an existing announcement is that the format of the edit URL is now different. Here are some samples from our test site:

Old Format (throws 404): [Root Site URL]/Home/tabid/73/ctl/Edit/itemid/325/mid=2044?popUp=true

New Format (works): [Root Site URL]/Home/tabid/73/ctl/Edit/mid/2044/itemid/325/Default.aspx?popUp=true

I don’t know if this should be reported as a separate issue or not – just letting you know.

Thanks,
Eric

From: Johnson, Eric
Sent: Thursday, July 5, 2018 10:44 AM
To: DNNCommunity/DNN.Announcements <reply@reply.github.com>; DNNCommunity/DNN.Announcements <DNN.Announcements@noreply.github.com>
Cc: Mention <mention@noreply.github.com>
Subject: RE: [DNNCommunity/DNN.Announcements] DotNetNuke.Modules.Announcements.Components.Business.AnnouncementInfo.GetProperty Error (#17)

Daniel,

Ignore what I said below – I found the issue. The “token replacement” references in the stack trace got me to thinking about how I’ve had to deal with NULL database values in some of our .NET apps. The “ImageSource” field in the Announcements table was NULL for several of our existing announcements, but adding a new announcement inserts an empty string for this field if you don’t use an image – this is why the new instance I mentioned below didn’t cause the error. I imagine a code change was made that assumes this value will never be null.

In the short term, we can fix this by simply updating our existing announcement records.

Thanks for your help,
Eric

From: Johnson, Eric
Sent: Thursday, July 5, 2018 9:15 AM
To: 'DNNCommunity/DNN.Announcements' <reply@reply.github.com>; DNNCommunity/DNN.Announcements <DNN.Announcements@noreply.github.com>
Cc: Mention <mention@noreply.github.com>
Subject: RE: [DNNCommunity/DNN.Announcements] DotNetNuke.Modules.Announcements.Components.Business.AnnouncementInfo.GetProperty Error (#17)

Daniel,

I can’t edit the module to even get to the template configuration – that is where the error occurs. I looked in the ModuleSettings table, though, and the template records appear to be in order.

I think the problem is that the new version cannot find the templates of existing announcement module instances – that is the reason for the null object reference. When I add a new announcements module instance, everything works fine.

Has the location of the settings moved? The reason I ask, is because when I created a new module instance, nothing was added to the ModuleSettings table.

I also cleared the cache during this testing.

Thanks,
Eric
